RenderingNG architecture post (#1013)
GoogleChrome/developer.chrome.com: site/en/blog/renderingng-architecture/index.md

---
layout: 'layouts/blog-post.njk'
title: 'Overview of the RenderingNG architecture'
description: >
  This post explains the components of the RenderingNG architecture,
  and how the rendering pipeline flows through them.
date: 2021-07-20
authors:
  - chrishtr
---

In a [previous post](/blog/renderingng/),
I gave an overview of the RenderingNG architecture goals and key properties.
This post will explain how its component pieces are set up,
and how the rendering pipeline flows through them.

Starting at the highest level and drilling down from there,
the tasks of rendering are:

1. Render contents into pixels on the screen.
1. Animate visual effects on the contents from one state to another.
1. Scroll in response to input.
1. Route input efficiently to the right places so that developer scripts and other subsystems can respond.

> The contents to render are a stream of raw input events from touch screens,

Author: I was trying to avoid the nested lists, which didn't really seem required here. The next paragraph then moved onto talking about frames, which seemed a more natural follow-on from "Each frame includes:" though I can swap them around, I think it would be good to avoid the nested list there.

> The contents to render are a stream of raw input events from touch screens,

Reviewer: Avoiding the nested list seems fine, but I think the order needs to be swapped.

mice, keyboards, and other hardware devices.
And, a tree of frames for each browser tab, plus the browser UI.
Each frame includes:

1. DOM state
1. CSS
1. Canvases
1. External resources, such as images, video, fonts, and SVG

> A frame is an HTML document, plus its URL.

Reviewer: LGTM. Also italicize "contents".

I also like changing bold to italics in all the other places.

(In reality,
[always on desktop; when possible on mobile](https://blog.chromium.org/2019/10/recent-site-isolation-improvements.html).
I'll write "always" below,
but this caveat applies throughout.)

Multiple browser tabs or windows of the same site usually go in different render processes,
unless the tabs are related (one [opening](https://developer.mozilla.org/en-US/docs/Web/API/Window/opener)
the other).
Under strong memory pressure on desktop Chromium may put multiple tabs
from the same site into the same render process even if not related.

Within a single browser tab,
frames from different sites are always in different render processes from each other,
but frames from the same site are always in the same render process.
From the perspective of rendering,
the important advantage of multiple render processes is that cross-site iframes
and tabs achieve [performance isolation](/blog/renderingng/#performance-isolation)
from each other.
In addition, origins can opt into [even more isolation](https://web.dev/origin-agent-cluster/).

There is exactly one Viz process for all of Chromium.

> After all, there is usually only one GPU and screen to draw to.

Author: It seemed unnecessary. That may be due to me being British, but the style guide agrees with me https://developers.google.com/style/exclamation-points

> After all, there is usually only one GPU and screen to draw to.

Reviewer: I wanted it to indicate personality/emotion/excitement, because this is not just code documentation but a blog post. WDYT, is there a requirement somewhere that developers.chrome.com downplay the voice of the writer more so than web.dev or the chromium blog?

Raster and draw generally happen on the same thread,
because both of them rely on GPU resources,
and it's hard to reliably make multi-threaded use of the GPU
(easier multi-threaded access to the GPU is one motivation for developing the new
[Vulkan](https://www.vulkan.org/) standard).
On Android WebView, there is a separate OS-level render thread for drawing
because of how WebViews are embedded into a native app.
Other platforms will likely have such a thread in the future.

The display compositor is on a different thread because it needs to be responsive at all times,
and not block on any possible source of slowdown on the GPU main thread.
One cause of slowdown on the GPU main thread is calls into non-Chromium code,
such as vendor-specific GPU drivers, that may be slow in hard-to-predict ways.

## Component structure

### Render process main thread components

{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/5pqtEVTK9xbbB7sc2s87.jpeg",
alt="Blink renderer.", width="800", height="464" %}

- **Blink renderer:**
  - **Local frame tree:** represents the tree of local frames and the DOM within frames.
  - **DOM and Canvas APIs:** implements of all of these APIs.

>   - **Document lifecycle runner:** executes the rendering pipeline steps up to and including the commit step.

Reviewer: ```suggestion
  - The **Document lifecycle runner** executes the rendering pipeline steps up to and including the commit step.
```

>   - **Input event hit testing and dispatching:** executes hit tests to find out which

Reviewer: ```suggestion
  - The **input event hit testing and dispatching** component executes hit tests to find out which
```

  DOM element is targeted by an event,
  and runs the input event dispatching algorithms and default behaviors.

> - **Rendering event loop scheduler and runner:** decides what to run on the event loop and when.

Reviewer: ```suggestion
- The **rendering event loop scheduler and runner** decides what to run on the event loop and when.
```

It schedules rendering to happen at a cadence matching the device display.

Local frame trees are a bit complicated to think about.
Recall that a frame tree is the main page and its child iframes, recursively.
A frame is local to a render process if it is rendered in that process,
and otherwise it is remote.

You can imagine coloring frames according to their render process.

> In the next example, the green circles are all frames in one render process;

Reviewer: ```suggestion
In the example above, the green circles are all frames in one render process;
```

Reviewer: (and move the image up, as we discussed)

the red ones are in a second, and the blue one is in a third.

A local frame tree is a connected component of the same color in a frame tree.
There are four local frame trees in the next example: two for site A, one for site B, and one for site C.
Each local frame tree gets its own Blink renderer component.
A local frame tree's Blink renderer may or may not be in the same render process as other local frame trees
(it's determined by the way render processes are selected, as described earlier).

{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/RKXD237nuR9aHxw6O2Mc.jpeg",
alt="Frame tree", width="800", height="629" %}

### Render process compositor thread structure

{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/zdmgrfrspiED4fOQtoca.jpeg",
alt="Render process compositor components.", width="800", height="559" %}

> Render process compositor components include:

Reviewer: ```suggestion
The render process compositor components include:
```

> - **Data handler:** maintains a composited layer list, display lists and property trees.

Reviewer: ```suggestion
- A **data handler** that maintains a composited layer list, display lists and property trees.
```

> - **Lifecycle runner:** runs the animate, scroll, composite, raster,

Reviewer: ```suggestion
- A **lifecycle runner** that runs the animate, scroll, composite, raster,
```

and decode and activate steps of the rendering pipeline.
(Remember that animate and scroll can occur in both the main thread and the compositor.)

> - **Input and hit test handler:** performs input processing and hit testing at the resolution of composited layers,

Reviewer: ```suggestion
- An **input and hit test handler** performs input processing and hit testing at the resolution of composited layers,
```

to determine if scrolling gestures can be run on the compositor thread,
and which render process hit tests should target.

## An example in practice

Let's now make the architecture concrete with an example.
In this example there are three tabs:

**Tab 1: foo.com**

```html
<html>

>   <iframe #one src="foo.com/other-url"></iframe>

Reviewer: ```suggestion
  <iframe id=one src="foo.com/other-url"></iframe>
```

>   <iframe  #two src="bar.com"></iframe>

Reviewer: ```suggestion
  <iframe  id=two src="bar.com"></iframe>
```

1. The event is sent, if necessary, to the main thread.
1. The main thread fires input event listeners
(pointerdown, touchstar, pointermove, touchmove or wheel)
to see if listeners will call preventDefault the event.
1. The main thread returns whether preventDefault was called to the compositor.
1. If not, the input event is sent back to the browser process.
1. The browser process converts it to a scroll gesture by combining it with other recent events.
1. The scroll gesture is sent once again to baz.com's render process compositor thread,
1. The scroll is applied there, and the compositor thread for the bar.com
render process ticks an animation in its compositor event loop.
This then mutates scroll offset in the property trees and re-runs the compositor lifecycle.
It also tells the main thread to fire a scroll event (not depicted here).
1. A compositor frame is sent to Viz.
1. Viz aggregates the compositor frames for the foo.com render process,
the bar.com render process, and the browser UI.
1. Viz schedules a draw.
1. Viz draws the aggregated compositor frame to the screen.

{% Aside %}
Note that each input event after the first can skip steps three and four,

> because the scroll has already begun and at that point can observe it,

Reviewer: ```suggestion
because the scroll has already begun and at that point scripts can observe it via the `scroll` event,
```

but no longer interrupt.
{% endAside %}

To route a click event on a hyperlink in iframe #two on tab one:

{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/GnXCsuXJkNegGDMQAhnx.jpeg",
alt="", width="800", height="488" %}

1. An input event (mouse, touch or keyboard) comes to the browser process.
It performs an approximate hit test

> to determine that the bar.com render process should receive the click, and sends it there.

Reviewer: ```suggestion
to determine that the bar.com iframe render process should receive the click, and sends it there.
```

1. The compositor thread for bar.com routes the click event to the main thread
for bar.com and schedules a rendering event loop task to process it.
1. The input event processor for bar.com's main thread hit tests to determine which
DOM element in the iframe was clicked, and fires a click event for scripts to observe.
Hearing no preventDefault, it navigates to the hyperlink.

> 1. Upon load, the new state is rendered, with steps similar to the "render changed DOM" example above.

Reviewer: ```suggestion
1. Upon load of destination page of the hyperlink, the new state is rendered, with steps similar to the "render changed DOM" example above.
```

(These subsequent changes are not depicted here.)

## Conclusion

Phew, that was a lot of detail.

> As you can see, rendering in Chromium is quite complicated.

Reviewer: ```suggestion
As you can see, rendering in Chromium is quite complicated!
```

It can take a lot of time to remember and internalize all the pieces,
so don't be worried if it seems overwhelming.

The most important takeaway is that there is a conceptually simple rendering pipeline,
which through careful modularization and attention to detail,
has been split into a number of self-contained components.
These components have then been split across parallel processes and threads to maximize
[scalable performance](/blog/renderingng/#scalable-performance)
and [extensibility](/blog/renderingng/#extensibility-the-right-tools-for-the-job) opportunities.

Each of those components plays a critical role in enabling
all of the performance and features modern web apps need.

> Next we'll publish deep-dives into each of them,

Reviewer: ```suggestion
Soon we'll publish deep-dives into each of them,
```

> and the important role they play.

Reviewer: ```suggestion
and the important roles they play.
```

> But before that, I'll also explain how the key data structures mentioned in this post

Reviewer: ```suggestion
But before that, next up I'll explain how the key data structures mentioned in this post
```

Author: I wondered about "But before that, the next post will explain ..." Before that, next up is a bit of an odd construction.

Reviewer: Maybe:

"Soon we'll publish deep-dives..."

And leave this sentence alone.

---
layout: 'layouts/blog-post.njk'
title: 'Overview of the RenderingNG architecture'
description: >
  This post explains the components of the RenderingNG architecture,
  and how the rendering pipeline flows through them.

> date: 2021-07-21

Author: ```suggestion
date: 2021-07-26
```

authors:
  - chrishtr
---

In a [previous post](/blog/renderingng/),
I gave an overview of the RenderingNG architecture goals and key properties.
This post will explain how its component pieces are set up,
and how the rendering pipeline flows through them.

Starting at the highest level and drilling down from there,
the tasks of rendering are:

1. _Render contents_ into pixels on the screen.
1. _Animate visual effects_ on the contents from one state to another.
1. _Scroll_ in response to input.
1. _Route input_ efficiently to the right places so that developer scripts and other subsystems can respond.

The _contents_ to render are a tree of _frames_ for each browser tab, plus the browser UI.
And, a stream of raw input events from touch screens,
mice, keyboards, and other hardware devices.

Each frame includes:

> 1. DOM state
> 1. CSS
> 1. Canvases
> 1. External resources, such as images, video, fonts, and SVG

Reviewer 2: ```suggestion
- DOM state
- CSS
- Canvases
- External resources, such as images, video, fonts, and SVG
```

The components end up in various CPU processes, threads,
and sub-components within those threads.
Each plays an important role in achieving
[reliability](/blog/renderingng/#reliability),
[scalable performance](/blog/renderingng/#scalable-performance)
and [extensibility](/blog/renderingng/#extensibility-the-right-tools-for-the-job) for all web content.

### Rendering pipeline structure

{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/jsWIjoou2iaN6ORvMfvM.jpeg",
alt="Diagram of the rendering pipeline as explained in the following text.",
width="800", height="1273" %}

Rendering proceeds in a pipeline with a number of stages and artifacts created along the way.
Each _stage_ represents code that does one well-defined task within rendering.
The _artifacts_ are data structures that are inputs or outputs of the stages;
in the diagram inputs or outputs are indicated with arrows.

This blog post won't go into much detail about the artifacts;
that will be discussed in the next post:
_Key Data Structures and their roles in RenderingNG_.

### The pipeline stages

In the preceeding diagram, stages are notated with colors indicating in which thread or process they execute:

> - Render process main thread: green
> - Render process compositor: yellow
> - Viz process: orange

Reviewer 2: ```suggestion
- **Green:** render process main thread
- **Yellow:** render process compositor
- **Orange:** viz process
```

Reviewer 2: I think this formatting makes the text easier to process and it's visually better balanced

Stages of the rendering pipeline can be skipped if they aren't needed.
For example, animations of visual effects, and scroll, can skip layout, pre-paint and paint.
This is why animation and scroll are marked with yellow and green dots in the diagram.
If layout, pre-paint, and paint can be skipped for visual effects,
they can be run entirely on the compositor thread and skip the main thread.

Browser UI rendering is not depicted directly here,
but can be thought of as a simplified version of this same pipeline
(and in fact its implementation shares much of the code).
Video (also not directly depicted)
generally renders via independent code that decodes frames into GPU texture tiles
that are then plugged into compositor frames and the draw step.

## Process and thread structure

{% Aside %}
There are more CPU processes and threads in the Chromium architecture than pictured here.
These diagrams focus only on those critical to rendering.
{% endAside %}

### CPU processes

The use of multiple CPU processes achieves performance and security isolation
between sites and from browser state,
and stability and security isolation from GPU hardware.

> {% Img

Reviewer 2: Consider using some crop of this as a header image, it will make the post stand out when shared on social media.

Author: last time the author decided not to go with a header image, @chrishtr let me know what you think.

Reviewer: Let's skip it.

On WebView, the browser process is also shared with the embedding app,
and WebView has only one render process.

There is also sometimes a utility process for decoding protected video content.
This process is not depicted above.

## Threads

Threads help achieve performance isolation and responsiveness in spite of slow tasks,
pipeline parallelization and multiple buffering.

{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/rdgwp4YLNUeUyuc6Eld1.jpeg", alt="A diagram of the render process as described in the article.",
width="800", height="556" %}

- The _main thread_ runs scripts, the rendering event loop, the document lifecycle,
hit testing, script event dispatching, and parsing of HTML, CSS and other data formats.
  - _Main thread helpers_ perform tasks such as creating image bitmaps and blobs that require encoding or decoding.
  - _[Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)_
  run script, and a rendering event loop for OffscreenCanvas.
- The _Compositor thread_ processes input events,
performs scrolling and animations of web content,
computes optimal layerization of web content,
and coordinates image decodes, paint worklets and raster tasks.

>   - _Compositor thread helpers:_ coordinate Viz raster tasks,

Reviewer: ```suggestion
  - _Compositor thread helpers_ coordinate Viz raster tasks,
```

To _animate_ a CSS transform transition on tab two:

{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/JBZ4Fo0wpEznvZIaz2wS.jpeg",
alt="", width="800", height="488" %}

1. The compositor thread for the bar.com render process ticks an animation
in its compositor event loop by mutating the existing property trees.
This then re-runs the compositor lifecycle. (Raster and decode tasks may occur, but are not depicted here.)
1. A compositor frame is sent to Viz.
1. Viz aggregates the compositor frames for the foo.com render process, the bar.com render process, and the browser UI.
1. Viz schedules a draw.
1. Viz draws the aggregated compositor frame to the screen.

To _scroll_ the web page on tab three:

{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/BTQMujxRdySU0rsgydwB.jpeg",
alt="", width="800", height="456" %}

1. A sequence of input events (mouse, touch or keyboard) come to the browser process.
1. Each event is routed to baz.com's render process compositor thread.
1. The compositor determines if the main thread needs to know about the event.
1. The event is sent, if necessary, to the main thread.
1. The main thread fires input event listeners

> (pointerdown, touchstar, pointermove, touchmove or wheel)

Author: ```suggestion
(`pointerdown`, `touchstar`, `pointermove`, `touchmove` or `wheel`)
```

> to see if listeners will call preventDefault the event.
> 1. The main thread returns whether preventDefault was called to the compositor.

Author: ```suggestion
to see if listeners will call `preventDefault` on the event.
1. The main thread returns whether `preventDefault` was called to the compositor.
```

1. If not, the input event is sent back to the browser process.
1. The browser process converts it to a scroll gesture by combining it with other recent events.
1. The scroll gesture is sent once again to baz.com's render process compositor thread,
1. The scroll is applied there, and the compositor thread for the bar.com
render process ticks an animation in its compositor event loop.
This then mutates scroll offset in the property trees and re-runs the compositor lifecycle.

> It also tells the main thread to fire a scroll event (not depicted here).

Author: ```suggestion
It also tells the main thread to fire a `scroll` event (not depicted here).
```

such as vendor-specific GPU drivers, that may be slow in hard-to-predict ways.

## Component structure
Within each render process main or compositor thread, there are logical software components that interact with each other in structured ways.
### Render process main thread components

{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/5pqtEVTK9xbbB7sc2s87.jpeg",
alt="Blink renderer.", width="800", height="464" %}

- **Blink renderer:**

>   - The **local frame tree** represents the tree of local frames and the DOM within frames.

Reviewer: ```suggestion
  - A **local frame tree fragment** represents the tree of local frames and the DOM within frames.
```

Reviewer: We decided in a draft of another post to use "fragment" in the name.

  - The **DOM and Canvas APIs** component contains implementations of all of these APIs.
  - The **Document lifecycle runner** executes the rendering pipeline steps up to and including the commit step.
  - The **input event hit testing and dispatching** component executes hit tests to find out which
  DOM element is targeted by an event,
  and runs the input event dispatching algorithms and default behaviors.
- The **rendering event loop scheduler and runner** decides what to run on the event loop and when.
It schedules rendering to happen at a cadence matching the device display.

> Local frame trees are a bit complicated to think about.

Reviewer: ```suggestion
Local frame trees fragments are a bit complicated to think about.
```

and otherwise it is remote.

You can imagine coloring frames according to their render process.
In the preceding example, the green circles are all frames in one render process;
the red ones are in a second, and the blue one is in a third.

> A local frame tree is a connected component of the same color in a frame tree.

Reviewer: ```suggestion
A local frame tree fragment is a connected component of the same color in a frame tree.
```
